plugins/crypto/ua_securitypolicy_filestore.c

```diff
@@ -0,0 +1,411 @@
+
+    char newCertFilename[PATH_MAX];
+    if(mp_snprintf(newCertFilename, PATH_MAX, "%s/%s%s", ownCertPathDir, newFilename, ".der") < 0)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    char newKeyFilename[PATH_MAX];
+    if(mp_snprintf(newKeyFilename, PATH_MAX, "%s/%s%s", ownKeyPathDir, newFilename, ".key") < 0)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    retval = writeByteStringToFile(newCertFilename, &newCertificate);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    return writeByteStringToFile(newKeyFilename, &newPrivateKey);
+}
+
+static UA_StatusCode
+updateCertificateAndPrivateKey_sp_filestore(UA_SecurityPolicy *securityPolicy,
+                                            const UA_ByteString newCertificate,
+                                            const UA_ByteString newPrivateKey) {
+    /* Temporarily save the old certificate file so that it can be removed from CertStore */
+    UA_ByteString localCertificateTmp;
+    UA_ByteString_copy(&securityPolicy->localCertificate, &localCertificateTmp);
+
+    SecurityPolicy_FilestoreContext *pc =
+        (SecurityPolicy_FilestoreContext *) securityPolicy->policyContext;
+
+    UA_StatusCode retval =
+        pc->policy.updateCertificateAndPrivateKey(securityPolicy, newCertificate, newPrivateKey);
+    if(retval != UA_STATUSCODE_GOOD) {
+        return retval;
+    }
+
+    retval =
+        writeCertificateAndPrivateKeyToFilestore(pc->storePath, newCertificate, newPrivateKey);
+    UA_ByteString_clear(&localCertificateTmp);
+
+    return retval;
+}
+
+static void
+clear_sp_filestore(UA_SecurityPolicy *securityPolicy) {
+    if(!securityPolicy)
+        return;
+
+    if(!securityPolicy->policyContext)
+        return;
+
+    SecurityPolicy_FilestoreContext *pc =
+            (SecurityPolicy_FilestoreContext *) securityPolicy->policyContext;
+
+    pc->policy.clear(&pc->policy);
+    UA_String_clear(&pc->storePath);
+
+    UA_free(pc);
+    securityPolicy->policyContext = NULL;
+}
+
+static void
+init_sp_filestore(UA_SecurityPolicy *policy) {
+    SecurityPolicy_FilestoreContext *pc =
+            (SecurityPolicy_FilestoreContext *) policy->policyContext;
+
+    policy->policyUri = pc->policy.policyUri;
+    policy->securityLevel = pc->policy.securityLevel;
+    policy->localCertificate = pc->policy.localCertificate;
+    policy->certificateGroupId = pc->policy.certificateGroupId;
+    policy->certificateTypeId = pc->policy.certificateTypeId;
+    policy->asymmetricModule = pc->policy.asymmetricModule;
+    policy->symmetricModule = pc->policy.symmetricModule;
+    policy->certificateSigningAlgorithm = pc->policy.certificateSigningAlgorithm;
+    policy->channelModule = pc->policy.channelModule;
+    policy->logger = pc->policy.logger;
+    policy->createSigningRequest = pc->policy.createSigningRequest;
+    policy->updateCertificateAndPrivateKey = updateCertificateAndPrivateKey_sp_filestore;
+    policy->clear = clear_sp_filestore;
```

Comment:
This looks like a complete clone of the policy within the context. Would it be better to just have the context pointing to the policy (instead of owning a duplicate) to avoid doing this copying stuff?